src/app/shared/pipes/filter.pipe.ts

```diff
@@ -0,0 +1,38 @@
+import { Pipe, PipeTransform } from "@angular/core";
+
+import { FilterType } from "@shared/models/types/filter-type.type";
+
+import { Product } from "@shared/models/classes/product.class";
+import { Coffee } from "@shared/models/classes/coffee.class";
+
+@Pipe({
+    name: "filter"
+})
+export class FilterPipe implements PipeTransform {
+
+    transform(products: Product[], filter: FilterType): Product[] {
+        let filteredProducts = [];
+
+        const coffees = products
```

Comment:
I know you had a comment about this, but the pipe should not be aware of the type of data structure it is receiving.

it should receive an arbitrary array of objects (object[]), along with an activeFilter (it is correct the way you have done it).

Then, we should (using the array.filter method):
1) loop through the keys of the object, check if the current key is included in the keys of the array of activeFilters
2) if it is found, it should check whether the value at that key in included in the activeFilters array at that same key (the value should check for a string or for an array of strings)

see below an example for a pipe that filters based on a searchString. The concept is similar, we just have to extend it a bit (note the first two lines)
`
import { Pipe, PipeTransform } from "@angular/core";

@Pipe({
    name: "filter"
})
export class FilterPipe implements PipeTransform {

    transform(items: object[], searchText: string): object[] {
        if (items == null) { return []; }
        if (searchText == null) { return items; }

        searchText = searchText.toLowerCase();

        return items.filter(item =>
            Object.keys(item).some((key) => {
                return typeof item[key] === "string" && item[key].toLowerCase().includes(searchText);
            })
        );
    }
}
`